ping/ping_common.c

```diff
@@ -806,35 +813,51 @@ int gather_statistics(struct ping_rts *rts, uint8_t *icmph, int icmplen,
 		uint8_t *cp, *dp;
 
 		print_timestamp(rts);
-		printf(_("%d bytes from %s:"), cc, from);
+		if (!rts->opt_json)
+			printf(_("%d bytes from %s:"), cc, from);
+		construct_json(rts, PING_JSON_INT, "bytes", cc);
+		// TODO: split host IP address and name into separate attributes in JSON?
+		construct_json(rts, PING_JSON_STR, "host", from);
 
 		if (pr_reply)
-			pr_reply(icmph, cc);
+			pr_reply(rts, icmph, cc);
 
 		if (rts->opt_verbose && rts->ident != -1)
-			printf(_(" ident=%d"), ntohs(rts->ident));
+			if (!rts->opt_json)
+				printf(_(" ident=%d"), ntohs(rts->ident));
 
-		if (hops >= 0)
-			printf(_(" ttl=%d"), hops);
+		if (hops >= 0) {
+			if (!rts->opt_json)
+				printf(_(" ttl=%d"), hops);
+
+			construct_json(rts, PING_JSON_INT, "ttl", hops);
+		}
 
 		if (cc < rts->datalen + 8) {
 			printf(_(" (truncated)\n"));
 			return 1;
 		}
 		if (rts->timing) {
+			char outtime[30];
+
 			if (rts->opt_rtt_precision)
-				printf(_(" time=%ld.%03ld ms"), triptime / 1000, triptime % 1000);
+				sprintf(outtime, "%ld.%03ld", triptime / 1000, triptime % 1000);
 			else if (triptime >= 100000 - 50)
-				printf(_(" time=%ld ms"), (triptime + 500) / 1000);
+				sprintf(outtime, "%ld ms", (triptime + 500) / 1000);
 			else if (triptime >= 10000 - 5)
-				printf(_(" time=%ld.%01ld ms"), (triptime + 50) / 1000,
+				sprintf(outtime, "%ld.%01ld", (triptime + 50) / 1000,
 				       ((triptime + 50) % 1000) / 100);
 			else if (triptime >= 1000)
-				printf(_(" time=%ld.%02ld ms"), (triptime + 5) / 1000,
+				sprintf(outtime, "%ld.%02ld", (triptime + 5) / 1000,
 				       ((triptime + 5) % 1000) / 10);
 			else
-				printf(_(" time=%ld.%03ld ms"), triptime / 1000,
+				sprintf(outtime, "%ld.%03ld", triptime / 1000,
 				       triptime % 1000);
+
+			if (!rts->opt_json)
+				printf(_(" time=%s ms"), outtime);
+			construct_json(rts, PING_JSON_STR, "time", outtime);
```

Comment:
I somehow thought that storing format in string is better than the time:
https://github.com/iputils/iputils/pull/574

But that way here would need another `sprintf()` so maybe it's not a good idea, maybe I should drop that PR.

Reply:
On a short glance, your patch looks cleaner than my multitude of `sprintf()` calls.